src/k8s-engine/wsl.ts

```diff
@@ -1137,73 +1152,106 @@ export default class WSLBackend extends events.EventEmitter implements K8s.Kuber
         this.lastCommandComment = 'Mounting WSL data';
         await this.progressTracker.action(this.lastCommandComment, 100, this.mountData());
         this.lastCommandComment = 'Starting WSL environment';
-        await Promise.all([
+        const installerActions = [
           this.progressTracker.action(this.lastCommandComment, 100, async() => {
             const logPath = await this.wslify(paths.logs);
-            const rotateConf = LOGROTATE_K3S_SCRIPT.replace(/\r/g, '').replace('/var/log', logPath);
 
-            await this.writeFile('/etc/init.d/k3s', SERVICE_SCRIPT_K3S, 0o755);
-            await this.writeFile('/etc/init.d/docker', SERVICE_SCRIPT_DOCKERD, 0o755);
-            await this.writeConf('docker', {
-              WSL_HELPER_BINARY: await this.getWSLHelperPath(),
-              LOG_DIR:           logPath,
-            });
-            await this.writeFile('/etc/logrotate.d/k3s', rotateConf, 0o644);
+            if (enabledK3s) {
+              const rotateConf = LOGROTATE_K3S_SCRIPT.replace(/\r/g, '')
+                .replace('/var/log', logPath);
+
+              await this.writeFile('/etc/init.d/k3s', SERVICE_SCRIPT_K3S, 0o755);
+              await this.writeFile('/etc/logrotate.d/k3s', rotateConf, 0o644);
+            }
+            if (this.#currentContainerEngine === ContainerEngine.CONTAINERD) {
+              await this.execCommand('mkdir', '-p', '/etc/cni/net.d');
+              await this.writeFile('/etc/cni/net.d/10-flannel.conflist', FLANNEL_CONFLIST, 0o644);
+              await this.writeFile('/etc/containerd/config.toml', CONTAINERD_CONFIG, 0o644);
+              await this.writeConf('containerd',
+                { log_owner: 'root', containerd_opts: `"--address=${ CONTAINERD_ADDRESS_K3S }"` });
+            } else if (this.#currentContainerEngine === ContainerEngine.MOBY) {
+              await this.writeFile('/etc/init.d/docker', SERVICE_SCRIPT_DOCKERD, 0o755);
+              await this.writeConf('docker', {
+                WSL_HELPER_BINARY: await this.getWSLHelperPath(),
+                LOG_DIR:           logPath,
+              });
+            }
+            await this.runInit();
             await this.writeFile(`/etc/init.d/buildkitd`, SERVICE_BUILDKITD_INIT, 0o755);
             await this.writeFile(`/etc/conf.d/buildkitd`, SERVICE_BUILDKITD_CONF, 0o644);
-            await this.runInit();
+            if (this.#currentContainerEngine === ContainerEngine.MOBY) {
+              if (!this.#enabledK3s) {
+                // k3s still requires docker if we're in moby mode...
+                await this.startService('docker', undefined);
+              }
+            } else {
+              // ... but we always need to run containerd explicitly.
+              await this.startService('containerd', undefined);
+            }
```

Comment:
The mismatch in the two args is confusing. We should either:
- Make `k3s` depend on either `docker` or `containerd`, depending on `$ENGINE`; don't manually start either when starting k3s
- Make `k3s` _not_ depend on either `docker` or `containerd`, and always start the engine manually.

Personally I favour option 1, but either way it should be consistent.

Reply:
Bump because comments were getting lost.

Reply:
> The mismatch in the two args is confusing. We should either:
> 
>     * Make `k3s` depend on either `docker` or `containerd`, depending on `$ENGINE`; don't manually start either when starting k3s
> 
>     * Make `k3s` _not_ depend on either `docker` or `containerd`, and always start the engine manually.
> 
> 
> Personally I favour option 1, but either way it should be consistent.

`wsl.ts` line 1177 ~ 1185.
>            if (this.#currentContainerEngine === ContainerEngine.MOBY) {
>              if (!this.#enabledK3s) {
>                // k3s still requires docker if we're in moby mode...
>                await this.startService('docker', undefined);
>              }
>            } else {
>              // ... but we always need to run containerd explicitly.
>              await this.startService('containerd', undefined);
>            }

Reply:
This has been done.